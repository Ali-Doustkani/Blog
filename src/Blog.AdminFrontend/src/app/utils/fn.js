const compose = (...functions) => args => functions.reduceRight((arg, fn) => fn(arg), args)

const dict = (...pairs) => key => {
   try {
      return pairs.find(pair => pair[0] === key)[1]
   } catch {
      throw new Error('Invalid key for pair')
   }
}

const ifElse = (predicate, onTrue, onFalse) => (state, action) =>
   predicate(action) ? onTrue(state, action) : onFalse(state, action)

export { compose, dict, ifElse }
