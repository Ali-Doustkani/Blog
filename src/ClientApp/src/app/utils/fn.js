Array.prototype.scalar = function() {
   return this.length === 1 ? this[0] : this
}

const compose = (...functions) => args => functions.reduceRight((arg, fn) => fn(arg), args)

const dict = (...pairs) => key => {
   try {
      return pairs.find(pair => pair[0] === key)[1]
   } catch {
      throw new Error('Invalid key for pair')
   }
}

const ifElse = (predicate, onTrue, onFalse) => (state, action) =>
   predicate(action) ? onTrue(state, action) : onFalse(state, action)

const addItem = (prop, getNew) => state => ({ ...state, [prop]: [...state[prop], getNew()] })

const updateItem = prop => (state, action) => ({
   ...state,
   [prop]: state[prop].map(obj =>
      action.change.id === obj.id ? { ...obj, ...action.change } : obj
   )
})

const deleteItem = prop => (state, action) => ({
   ...state,
   [prop]: state[prop].filter(x => x.id !== action.id)
})

function today() {
   const now = new Date()
   return `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, 0)}-${String(
      now.getDate()
   ).padStart(2, 0)}`
}

export { compose, dict, ifElse, addItem, updateItem, deleteItem, today }
