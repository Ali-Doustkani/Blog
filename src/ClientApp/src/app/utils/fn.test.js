import { dict, ifElse, addItem, updateItem, deleteItem } from './fn'

describe('dict', () => {
   const a = {},
      b = {},
      c = {},
      func = dict(['a', a], ['b', b], ['c', c])

   it('selects the correct func', () => {
      expect(func('a')).toBe(a)
      expect(func('b')).toBe(b)
      expect(func('c')).toBe(c)
   })

   it('throws when invalid key is passed', () => {
      expect(() => func('d')).toThrow('Invalid key for pair')
   })
})

describe('ifElse', () => {
   it('returns the first argument when true', () => {
      const first = (state, action) => 'first: ' + state.value + action.value
      const second = (state, action) => 'second: ' + state.value + action.value
      const fn = ifElse(() => true, first, second)
      const result = fn({ value: 'a' }, { value: 'b' })
      expect(result).toBe('first: ab')
   })

   it('returns the second argument when false', () => {
      const first = (state, action) => 'first: ' + state.value + action.value
      const second = (state, action) => 'second: ' + state.value + action.value
      const fn = ifElse(() => false, first, second)
      const result = fn({ value: 'a' }, { value: 'b' })
      expect(result).toBe('second: ab')
   })
})

describe('state manipulation', () => {
   it('adds a new item', () => {
      const init = { values: [] }
      const newState = addItem('values', () => ({ id: 1, name: 'Ali' }))(init)
      expect(newState).toEqual({ values: [{ id: 1, name: 'Ali' }] })
   })

   it('deletes an existing value', () => {
      const initial = {
         values: [{ id: 1 }, { id: 2 }]
      }
      const newState = deleteItem('values')(initial, { id: 1 })
      expect(newState).toEqual({ values: [{ id: 2 }] })
   })

   it('updates an item', () => {
      const init = {
         values: [{ id: 1, name: 'Ali', surname: 'Doustkani' }]
      }

      const update = updateItem('values')
      let state = update(init, { change: { id: 1, name: 'Haniye' } })
      state = update(state, { change: { id: 1, surname: 'Doosty' } })

      expect(state).toEqual({ values: [{ id: 1, name: 'Haniye', surname: 'Doosty' }] })
   })

   it('keeps the state intact', () => {
      let id = 1
      const getId = () => {
         id++
         return id
      }

      const initialState = { values: [] }
      const add = addItem('values', () => ({ id: getId(), name: '', surname: '' }))
      const update = updateItem('values')
      const deleteIt = deleteItem('values')

      let state = add(initialState)
      state = add(state)
      state = update(state, { change: { id: 2, name: 'Ali' } })
      state = update(state, { change: { id: 2, surname: 'Doustkani' } })
      state = deleteIt(state, { id: 2 })

      expect(state).toMatchSnapshot()
   })
})
