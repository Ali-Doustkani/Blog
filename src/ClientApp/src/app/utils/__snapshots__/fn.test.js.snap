// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`state manipulation keeps the state intact 1`] = `
Object {
  "values": Array [
    Object {
      "id": 3,
      "name": "",
      "surname": "",
    },
  ],
}
`;
